README.md

```diff
@@ -136,15 +136,14 @@ to the EC2 instance.
 
 * `security_groups` - (Optional) List of security group names.
 
-* `tags` - (Optional) A mapping of tags to assign to the resource.
+* `tags` - (Optional) A mapping of tags to assign where supported.
 
 * `template_file` - (Optional) User data template file.
 
-* `tier` - (Optional) A subnet tier tag (e.g., public, private,
-nat) to determine subnets to be associated with the load balancer.
-
 * `private_ip` - (Optional) Private IP address to associate with the instance in a VPC.
 
+* `use_private_ip - (Optional) Route53 record should use private rather than public IP address.
```

Comment:
@JonRoma , this setting should be an attribute of the alias block. If I'm reading this right, you have it documented as a top-level setting in the inputs block, not a parameter tha goes inside an alias block.

Reply:
@kwessel, I think you are confusing the top-level `use_private_ip` argument that we added with the `private_zone` attribute that's part of the `alias` block.

I suppose one could make an argument that the choice between the private *vs.* public IP could be on a per-alias basis, but I frankly can't think of a use case that compels using multiple Route53 aliases for a single EC2 instance, and such a use case would *only* be functional for a service deployed in the "campus" subnets.

Going a step farther, I'm not sure if @ddriddle and I had a specific use case in mind when we designed the module to support multiple Route53 aliases, or whether this feature was added to the module simply because we could.

Since I'm off on medical leave, I'm not going to merge this now, and hopefully we will hear from @edthedev's team about the question posed earlier in this PR's conversation.

Reply:
Nope, I wasn't talking about the private_ip setting. I was talking about the new use_private_ip that we added yesterday. I still think this is confusing. The hostname and domain as well as the Boolean private_zone setting all go in an alias block, yet what address to assign to that alias (the use_private_ip setting) goes outside that block and thus applies to all aliases. I can't think of a strong use case when an EC2 instance would need multiple aliases. It would be possible, given that the public and private IPs are assigned to different network interfaces on teh EC2 box, to have different things responding publically vs. privately. While I can't think of why someone wuld want to do that with an EC2 box, it's within the realm of possibilities, and in that case, you'd want two aliases: one being a hostname that resolves to the public IP and another hostname that resolves to the private IP. To do that, the use_private_ip, like the private_zone setting, woudl need to be inside the alias block and apply specifically to that alias.

If you don't wish to support this, then I'd encourage you to get rid of the alias block, assuming that's not a requirement of a downstream AWS module, and put hostname, domain, and private_zone as direct settings of the inputs block.

And holy cow, I can't believe you're up and alert and reading emails, Jon! Way to go, my friend!

Reply:
@kwessel, my response was relating to what you said:

> If I'm reading this right, you have it documented as a top-level setting in the inputs block, not a parameter that goes inside an alias block.

That's because the `use_private_ip` *is* defined in the top-level of the inputs block, unlike the other setting.

It would not be hard to move the new parameter into the `alias` block, and I am open to that idea. But I wanted to hear @ddriddle's views as well as those of @tzturner, who submitted the PR that added the new `private_ip` relatively recently.

The pertinent points to discuss are:

**When we originally coded this module, did we actually identify a use case that justifies having an `alias` block at all, or did we just do this because we could?**

I don't like supporting code that makes things needlessly complex if we gain nothing from that complexity. If we were starting anew, I would not implement multiple domain/hostname pairs as we did without good reason.

However, given the status quo, I'm not entirely eager to rip out working code implementing the `alias` block by placing the domain/hostname as top-level input variables. If we *really* wanted to gold plate it, the alias block could be a map containing domains (zones) as keys with a list of hostnames as values. :-)

**With the new capability we added yesterday, does SecDev still have a use case for the `private_ip` that was added, or does our new feature resolve that? I do not know what their needs are.

**We should think about the ambiguity in naming the arguments `private_ip`, `use_private_ip`, and the `private_zone` attribute of the `alias` block. This is potentially confusing to the wider audience we hope to reach, so now is a good time to consider naming or at least improving the documentation.

The reason I want more opinions on this is that I am on the fence.

Reply:
@kwessel wrote:

> And holy cow, I can't believe you're up and alert and reading emails, Jon! Way to go, my friend!

That's because my procedure doesn't take place till Friday. :-)

Reply:
@JonRoma I hope you procedure goes well and you feel well soon!

I agree with @kwessel that the new argument `use_private_ip` belongs in the alias block. At least from a UI perspective. 

> Going a step farther, I'm not sure if @ddriddle and I had a specific use case in mind when we designed the module to support multiple Route53 aliases, or whether this feature was added to the module simply because we could.

I had little to do with the development of this particular module. I'm not sure why you added support for multiple alias blocks but I suspect it is because we did this elsewhere and you copied code from another module, or maybe it is something that Maiko requested at the time. Whatever the initial reason, I think it is a good feature and worth keeping if the work to maintain it is not too onerous. In general, I prefer generic solutions to overly specific solutions unless there is a good reason such as maintenance cost or performance.

Reply:
@kwessel and @ddriddle, you've convinced me to get off the fence and make `use_private_ip` an attribute of `alias`. After thinking about it for a while, this makes sense because the type of IP address is tied to an ALIAS record.

However, this doesn't necessarily produce a winning combination. See the following snippet:

```
# Route53 records created for the EC2 instance.
  alias = [
    {
      hostname       = "identity-crisis"
      domain         = "test.iamu.techservices.illinois.edu"
      use_private_ip = true
    },
    {
      hostname       = "crisis-of-identity"
      domain         = "test.iamu.techservices.illinois.edu"
      use_private_ip = false
    }
  ]

  subnet_type                 = "campus"
  associate_public_ip_address = true
```
From outside campus IP space (i.e., when I am at home and **not** on the VPN), I can't SSH to the box using **either** of the aliases:

1. The private IP is inaccessible because campus won't route off-campus traffic into RFC1918 space.
1. The public IP is inaccessible because campus-or private-facing subnet doesn't route traffic from outside the VPC.

Using a private IP for a resource in the "public" `subnet_type` won't work either for reason 1. Likewise, using a public IP for a resource in the "private" `subnet_type` also fails for reason 2.

If I read the situation correctly, supporting multiple aliases is worthless unless **all** aliases have the same value for the `use_private_ip` boolean **and** are in the "campus" `subnet_type` **and** the would-be SSH user is on campus (which includes being on the VPN).

The support for multiple aliases now strikes me as creating confusion without providing meaningful value. I am beginning to think that the "flexibility" of supporting multiple aliases is a mirage because of network routing.

Rather than ditching the `alias` block, I wonder if it should instead become a simple map rather than a list of maps. This would continue to group the related arguments together without building false expectations. Here is what I suggest: 

```
# Route53 records created for the EC2 instance.
  alias = {
    hostname       = "identity-crisis"
    domain         = "test.iamu.techservices.illinois.edu"
    use_private_ip = true
  }
```

Thoughts?

Reply:
I'm also interested in feedback on [the questions posed in this comment](https://github.com/techservicesillinois/terraform-aws-ec2/pull/23#discussion_r1019734964). I'm looking to implement this as soon as we have an adequate discussion since we can nearly immediately deploy code using a public IP address in a campus-facing subnet.

Reply:
> I'm also interested in feedback on [the questions posed in this comment](https://github.com/techservicesillinois/terraform-aws-ec2/pull/23#discussion_r1019734964). I'm looking to implement this as soon as we have an adequate discussion since we can nearly immediately deploy code using a public IP address in a campus-facing subnet.

Reply:
> 2. The public IP is inaccessible because a campus-or private-facing subnet doesn't accept inbound traffic from outside the VPC.

Is that true just because of the routing in our campus-facing subnet? Or is that inherent to any type of campus-facing subnet?

I can understand this argument, but I can also see a campus-facing subnet that's designed to also accept inbound public traffic. If that's doable, then assigning both private and public IP addresses to hostnames could be useful. Perhaps this is way too much of a niche case and I'm overthinking this. However, the code's there and working, and the only issue is the current routing of the subnet in which it's deployed. In a different situation, this could work. My vote woudl be to leave it as-is, but I'm willing to go either way on this as you all feel fit.

Reply:
@kwessel wrote:

> > 2. The public IP is inaccessible because a campus-or private-facing subnet doesn't accept inbound traffic from outside the VPC.
> 
> Is that true just because of the routing in our campus-facing subnet? Or is that inherent to any type of campus-facing subnet?

Far be it from me to attempt to explain networking with any degree of competence, but the behavior is described by David Z's [Amazon Web Services, VPC Guide for Illinois](https://answers.uillinois.edu/illinois/page.php?id=71015#subnet-types).

In the table below reproduced from that document (omitting IPv6), here is a table that may or may not pass an accessibility test.

  | Public-facing subnet | Private-facing subnet | Campus-facing subnet
-- | -- | -- | --
To / from private IPv4 in same VPC | yes | yes | yes
To / from private IPv4 in other VPC | yes (via Transit Gateway or VPC peering) | yes (via Transit Gateway or VPC peering) | yes (via Transit Gateway or VPC peering)
Outbound IPv4 to Internet (or any publicly reachable server) | yes (requires public IP or Elastic IP) | yes (via NAT Gateway in your VPC) | yes (via NAT Gateway in your VPC or shared NAT Gateway in Core Services VPC)
Inbound IPv4 from Internet (including campus clients) to AWS public IP | yes (requires public IP or Elastic IP) | no | no
Outbound to campus host (not publicly reachable) from AWS private IPv4 | no | no | yes
Inbound from campus host to AWS private IPv4 | no | no | yes

Summing up the answer to your question, inbound IPv4 traffic from the Internet (including campus clients) is **not** routed to campus-facing and private-facing subnets. However, such inbound traffic *is* routed to a public-facing subnet, provided the destination address has a public IP or elastic IP.

David Z. went on to say that 
> Campus-facing subnets support bi-directional communication without NAT to campus networks at all three universities in the U of I System (some exceptions apply, see the Detailed Example).  They are otherwise similar to private-facing subnets, but currently support only IPv4 (not IPv6).

I interpret this as a statement that neither campus- nor private-facing subnets route traffic from the public Internet – and that, accordingly, you can only access an EC2 box placed in a campus-facing subnet from within the VPC or from a campus source address.

Reply:
Right, we're all already clear on how it works. That's how it's defined right now. That's not to say that additional public routing rules couldn't be added in the future. Clearly, nothing in the campus VPC would work with a public IP. Outbound traffic is NAT'd if it's going over the public internet, and inbound traffic won't work. So, to follow the argument you're making, @JonRoma , of saying that the current state of this code could lead to confusion would also argue against things like assign_public_ip in the campus facing subnet. I think the code, as it is right now, allows for maximum flexability, and it's up to the deployer to use smarts as to the routing rules of the VPC in which they're deploying. Should additional routing rules be added to campus-facing subnets in the future, it leaves room more more flexability with the code that's currently there. I don't think the current state of things hurts anything.